<a href="https://colab.research.google.com/github/JaCaRego/JaCaRego/blob/main/estatistica_mega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Perform a preliminary analysis of the Mega Sena results, including data loading from "https://raw.githubusercontent.com/carlosfab/data_science/master/datasets/mega_sena.csv" or "mega_sena.csv", data inspection and cleaning, data preparation, preliminary analysis, and visualization of the frequency of numbers, presenting key insights and a generated plot.

## Carregar o Arquivo Excel no Pandas

### Subtask:
Carregar os resultados da Mega Sena em um DataFrame pandas.


**Reasoning**:
First, I'll import the pandas library to work with DataFrames, then I'll load the CSV file from the provided URL into a pandas DataFrame named `df`.



# Task
Load the Mega Sena results from the "https://raw.githubusercontent.com/carlosfab/data_science/master/datasets/mega_sena.csv" CSV file into a pandas DataFrame.

## Carregar Arquivo Excel

### Subtask:
Carregar os resultados da Mega Sena em um DataFrame pandas.


**Reasoning**:
First, I'll import the pandas library to work with DataFrames, then I'll load the CSV file from the provided URL into a pandas DataFrame named `df`.



In [16]:
import pandas as pd

files.upload('Mega-Sena.xlsx')

df = pd.read_excel('mega_sena.xlsx')
print("DataFrame 'df' loaded successfully. Displaying the first 5 rows:")
df.head()

NameError: name 'files' is not defined

**Reasoning**:
The previous attempt to load the CSV from the URL resulted in an HTTP 404 error, indicating the online resource is unavailable. As per the task description, I will now try to load the CSV from the local file named `mega_sena.csv` using the `file_name` variable.



# Task
Confirm that the Excel file "Mega-Sena.xlsx" has been successfully loaded into a pandas DataFrame in the Colab environment and is ready for analysis.

## Fazer Upload do Arquivo Excel

### Subtask:
Confirm that the 'Mega-Sena.xlsx' file is available in the Colab environment. If not, upload it.


## Carregar o Arquivo Excel no Pandas

### Subtask:
Load the 'Mega-Sena.xlsx' file into a pandas DataFrame named `df` and display the first 5 rows to verify the loading.


**Reasoning**:
The subtask requires loading the 'Mega-Sena.xlsx' file into a pandas DataFrame and displaying its head. I will use `pd.read_excel` to load the file and `df.head()` to display the first 5 rows, as pandas is already imported.



In [17]:
df = pd.read_excel('Mega-Sena.xlsx')
print("DataFrame 'df' loaded successfully. Displaying the first 5 rows:")
df.head()

DataFrame 'df' loaded successfully. Displaying the first 5 rows:


Concurso Data do Sorteio  Bola1  Bola2  Bola3  Bola4  Bola5  Bola6  \
0         1      11/03/1996      4      5     30     33     41     52   
1         2      18/03/1996      9     37     39     41     43     49   
2         3      25/03/1996     10     11     29     30     36     47   
3         4      01/04/1996      1      5      6     27     42     59   
4         5      08/04/1996      1      2      6     16     19     46   

   Ganhadores 6 acertos Cidade / UF Rateio 6 acertos  Ganhadores 5 acertos  \
0                     0         NaN           R$0,00                    17   
1                     1          PR   R$2.307.162,23                    65   
2                     2      RN; SP     R$391.192,51                    62   
3                     0         NaN           R$0,00                    39   
4                     0         NaN           R$0,00                    98   

  Rateio 5 acertos  Ganhadores 4 acertos Rateio 4 acertos Acumulado 6 acertos  \
0      R$39.158,92                  2016         R$330,21      R$1.714.650,23   
1      R$14.424,02                  4488         R$208,91              R$0,00   
2      R$10.515,93                  4261         R$153,01              R$0,00   
3      R$15.322,24                  3311         R$180,48        R$717.080,75   
4       R$5.318,10                  5399          R$96,53      R$1.342.488,85   

  Arrecadação Total Estimativa prêmio  \
0            R$0,00            R$0,00   
1            R$0,00            R$0,00   
2            R$0,00            R$0,00   
3            R$0,00            R$0,00   
4            R$0,00            R$0,00   

  Acumulado Sorteio Especial Mega da Virada Observação  
0                                    R$0,00        NaN  
1                                    R$0,00        NaN  
2                                    R$0,00        NaN  
3                                    R$0,00        NaN  
4                                    R$0,00        NaN

## Gerar Jogos com Números de Maior Probabilidade

### Subtask:
Gerar 3 jogos, cada um composto por 6 números, selecionados entre os 11 números com maior frequência de sorteio na história da Mega Sena.

**Reasoning**:
First, I need to get the top 11 most frequent numbers from the `frequency_df`. Then, I will use Python's `random` module to select 6 unique numbers three times from this set of 11 to form the 3 games, ensuring each game is unique.

In [22]:
import random

# Get the top 11 most frequent numbers
top_11_numbers = frequency_df.nlargest(11, 'Frequency')['Number'].tolist()

print(f"Os 11 números mais frequentes são: {top_11_numbers}\n")

games = []
# Generate 3 unique games with 6 numbers from the top 11
while len(games) < 3:
    game = sorted(random.sample(top_11_numbers, 6))
    if game not in games:
        games.append(game)

print("Aqui estão os 3 jogos gerados com os números de maior probabilidade:")
for i, game in enumerate(games):
    print(f"Jogo {i+1}: {game}")

Os 11 números mais frequentes são: [10, 53, 34, 5, 37, 33, 38, 4, 17, 27, 32]

Aqui estão os 3 jogos gerados com os números de maior probabilidade:
Jogo 1: [4, 10, 17, 32, 37, 53]
Jogo 2: [4, 10, 27, 34, 37, 38]
Jogo 3: [4, 17, 27, 33, 34, 53]


## Final Task

### Subtask:
Confirm that 3 unique games, each with 6 numbers from the top 11 most frequent numbers, have been successfully generated and displayed.

## Final Task

### Subtask:
Confirm that the Excel file 'mega_sena.xlsx' has been successfully loaded into a pandas DataFrame in the Colab environment and is ready for analysis.


## Summary:

### Q&A
Yes, the Excel file 'Mega-Sena.xlsx' has been successfully loaded into a pandas DataFrame named `df` in the Colab environment and is ready for analysis.

### Data Analysis Key Findings
*   The 'Mega-Sena.xlsx' file was confirmed to be present in the Colab environment at `/content/Mega-Sena.xlsx`, eliminating the need for an upload.
*   The Excel file was successfully loaded into a pandas DataFrame named `df` using `pd.read_excel()`.
*   The initial rows of the DataFrame, displayed by `df.head()`, confirmed that the data was loaded correctly, including columns such as 'Concurso', 'Data do Sorteio', and 'Bola1' through 'Bola6'.

### Insights or Next Steps
*   The data is now prepared for further analysis, such as exploring descriptive statistics, identifying patterns in the drawn numbers, or preparing for machine learning models.


# Task
Select the columns from the DataFrame `df` that represent the drawn numbers (e.g., 'Bola1', 'Bola2', up to 'Bola6') and convert them to a numeric data type, if they are not already.

## Preparar Colunas de Números

### Subtask:
Selecionar as colunas do DataFrame `df` que contêm os números sorteados (Bola1, Bola2, etc.) e convertê-las para o formato numérico, se ainda não estiverem.


**Reasoning**:
I will create a list of the 'Bola' columns and then convert their data type to integer using `astype(int)` to ensure they are numeric for analysis.



In [18]:
bola_columns = ['Bola1', 'Bola2', 'Bola3', 'Bola4', 'Bola5', 'Bola6']
df[bola_columns] = df[bola_columns].astype(int)
print("Columns 'Bola1' through 'Bola6' converted to integer type. Displaying info:")
df[bola_columns].info()

Columns 'Bola1' through 'Bola6' converted to integer type. Displaying info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2946 entries, 0 to 2945
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Bola1   2946 non-null   int64
 1   Bola2   2946 non-null   int64
 2   Bola3   2946 non-null   int64
 3   Bola4   2946 non-null   int64
 4   Bola5   2946 non-null   int64
 5   Bola6   2946 non-null   int64
dtypes: int64(6)
memory usage: 138.2 KB


## Consolidar Números Sorteados

### Subtask:
Transformar as colunas de números sorteados do DataFrame `df` em uma única série ou lista de todos os números, para que possamos calcular a frequência de cada um.


**Reasoning**:
To consolidate the drawn numbers, I will select the 'Bola' columns, stack them into a single series, and store it in 'all_numbers' to prepare for frequency calculation.



In [19]:
all_numbers = df[bola_columns].stack()
print("Consolidated numbers (first 10):")
print(all_numbers.head(10))

Consolidated numbers (first 10):
0  Bola1     4
   Bola2     5
   Bola3    30
   Bola4    33
   Bola5    41
   Bola6    52
1  Bola1     9
   Bola2    37
   Bola3    39
   Bola4    41
dtype: int64


## Calcular Frequência dos Números

### Subtask:
Calcular a frequência de cada número na série consolidada. Isso nos dirá quantas vezes cada número foi sorteado na história da Mega Sena.


**Reasoning**:
To calculate the frequency of each number, I will use the `value_counts()` method on the `all_numbers` series and store the result in `frequency_counts`, then display the head of the resulting series.



In [20]:
frequency_counts = all_numbers.value_counts()
print("Frequency of each number (top 10 most frequent):")
print(frequency_counts.head(10))

Frequency of each number (top 10 most frequent):
10    343
53    335
34    320
5     320
37    317
33    316
38    315
4     312
17    312
27    312
Name: count, dtype: int64


## Armazenar Frequências para Análise

### Subtask:
Criar um novo DataFrame ou série a partir do DataFrame `df` para armazenar as frequências calculadas, com os números e suas respectivas contagens, facilitando análises futuras e visualizações.


**Reasoning**:
To store the calculated frequencies in a structured format, I will convert the `frequency_counts` series into a DataFrame, reset its index to make the numbers a column, and then rename the columns for clarity as 'Number' and 'Frequency'. Finally, I'll display the first 5 rows to verify the operation.



In [21]:
frequency_df = frequency_counts.to_frame().reset_index()
frequency_df.columns = ['Number', 'Frequency']
print("Frequency DataFrame created. Displaying the first 5 rows:")
print(frequency_df.head())

Frequency DataFrame created. Displaying the first 5 rows:
   Number  Frequency
0      10        343
1      53        335
2      34        320
3       5        320
4      37        317


## Final Task

### Subtask:
Confirm that the frequencies of the drawn numbers have been calculated and saved in an appropriate structure for analysis, and display a preview of this structure.


## Summary:

### Data Analysis Key Findings
*   The columns representing the drawn numbers ('Bola1' through 'Bola6') were successfully identified and converted to integer (`int64`) data type.
*   All drawn numbers across these columns were consolidated into a single pandas Series, facilitating aggregate analysis.
*   The frequency of each unique number was calculated: for instance, number 10 was drawn 343 times, 53 was drawn 335 times, and both 34 and 5 were drawn 320 times.
*   The calculated frequencies were structured into a new DataFrame named `frequency_df` with columns 'Number' and 'Frequency', ready for further analysis.

### Insights or Next Steps
*   The structured frequency data (e.g., `frequency_df`) is now suitable for various analyses, such as identifying the hottest (most frequent) and coldest (least frequent) numbers, which could be valuable for players or for understanding draw dynamics.
*   The next logical step would be to visualize these frequencies (e.g., using bar charts or heatmaps) to provide a clear and intuitive understanding of the distribution of drawn numbers over time.


# Task
Confirm that 3 unique games, each with 6 numbers from the top 11 most frequent numbers, have been successfully generated and displayed.

## Identificar os 11 Números para Base

### Subtask:
Confirm that the top 11 most frequent numbers have been identified and are stored in the `top_11_numbers` variable.


**Reasoning**:
To confirm that the `top_11_numbers` variable has been correctly identified and populated, I will display its content as requested by the instructions.



In [23]:
print(f"The top 11 most frequent numbers are: {top_11_numbers}")

The top 11 most frequent numbers are: [10, 53, 34, 5, 37, 33, 38, 4, 17, 27, 32]


## Final Task

### Subtask:
Confirm that 3 unique games, each with 6 numbers from the top 11 most frequent numbers, have been successfully generated and displayed.


## Summary:

### Data Analysis Key Findings
*   The `top_11_numbers` variable was successfully populated with the 11 most frequent numbers: `[10, 53, 34, 5, 37, 33, 38, 4, 17, 27, 32]`.
*   This confirms that the initial step of identifying the base set of numbers for game generation was completed successfully.

### Insights or Next Steps
*   The next step should focus on generating the 3 unique games using exactly 6 numbers from the confirmed `top_11_numbers` list.
*   Validate that the generated games are unique and correctly displayed as required by the final task.
